subsys/bluetooth/shell/bt.c

```diff
@@ -1147,14 +1160,83 @@ static int cmd_scan_filter_clear_addr(const struct shell *sh, size_t argc,
 #endif /* CONFIG_BT_OBSERVER */
 
 #if defined(CONFIG_BT_BROADCASTER)
-static const struct bt_data ad_discov[] = {
-	BT_DATA_BYTES(BT_DATA_FLAGS, (BT_LE_AD_GENERAL | BT_LE_AD_NO_BREDR)),
-};
+static void ad_mode_parse(const struct shell *sh, size_t argc, char *argv[], bool *discov)
+{
+	for (size_t argn = 0; argn < argc; argn++) {
+		const char *arg = argv[argn];
+
+		if (!strcmp(arg, "discov")) {
+			*discov = true;
+			break;
+		} else if (!strcmp(arg, "non_discov")) {
+			*discov = false;
+			break;
+		}
+	}
+}
+
+static ssize_t ad_init(struct bt_data *data, const size_t data_size, const atomic_t *adv_opt)
+{
+	const bool discoverable = atomic_test_bit(adv_opt, SHELL_ADV_OPT_DISCOVERABLE);
+	const bool adv_ext = atomic_test_bit(adv_opt, SHELL_ADV_OPT_EXT_ADV);
+	static uint8_t ad_flags;
+	size_t ad_len = 0;
+
+	/* Set BR/EDR Not Supported if LE-only device */
+	ad_flags = IS_ENABLED(CONFIG_BT_BREDR) ? 0 : BT_LE_AD_NO_BREDR;
+
+	if (discoverable) {
+		/* A privacy-enabled Set Member should advertise RSI values only when in
+		 * the GAP Limited Discoverable mode.
+		 */
+		if (IS_ENABLED(CONFIG_BT_PRIVACY) && IS_ENABLED(CONFIG_BT_CSIS) && csis != NULL) {
+			ad_flags |= BT_LE_AD_LIMITED;
+		} else {
+			ad_flags |= BT_LE_AD_GENERAL;
+		}
+	}
+
+	if (ad_flags != 0) {
+		__ASSERT(data_size > ad_len, "No space for AD_FLAGS");
+		data[ad_len].type = BT_DATA_FLAGS;
+		data[ad_len].data_len = sizeof(ad_flags);
+		data[ad_len].data = &ad_flags;
+		ad_len++;
+	}
+
+	if (IS_ENABLED(CONFIG_BT_CSIS)) {
+		ssize_t csis_ad_len;
+
+		csis_ad_len = csis_ad_data_add(&data[ad_len], data_size - ad_len, discoverable);
+		if (csis_ad_len < 0) {
+			shell_error(ctx_shell, "Failed to add CSIS data (err %d)", csis_ad_len);
+			return ad_len;
+		}
+
+		ad_len += csis_ad_len;
+	}
+
+	if (IS_ENABLED(CONFIG_BT_AUDIO) && IS_ENABLED(CONFIG_BT_EXT_ADV) && adv_ext) {
+		const bool connectable = atomic_test_bit(adv_opt, SHELL_ADV_OPT_CONNECTABLE);
+		size_t audio_ad_len;
+
+		audio_ad_len = audio_ad_data_add(&data[ad_len], data_size - ad_len,
+						 discoverable, connectable);
+		if (audio_ad_len < 0) {
+			return audio_ad_len;
+		}
+
+		ad_len += audio_ad_len;
+	}
+
+	return ad_len;
+}
 
 static int cmd_advertise(const struct shell *sh, size_t argc, char *argv[])
 {
 	struct bt_le_adv_param param = {};
-	const struct bt_data *ad;
+	struct bt_data ad[2];
```

Comment:
`2` seems like a very low number if we want to add flags, service UUIDs and multiple different service data entries.

Reply:
`cmd_advertise` starts legacy advertisement. Service UUIDs are advertised in extended advertisement only. The RSI is going to be added for legacy advertisement, thus we need 2 entries only (Flags + RSI)